patchwork-registries/src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -19,189 +19,804 @@
-			throw new IllegalArgumentException(String.format("The object %s{%x} has been registered twice, using the names %s and %s.", value, System.identityHashCode(value), oldIdentifier, identifier));
-		}
+	@Override
+	public Identifier getRegistryName() {
+		return this.name;
+	}
 
-		Registry.register(vanilla, identifier, value);
+	@Override
+	public Class<V> getRegistrySuperType() {
+		return superType;
 	}
 
 	@Override
-	public void registerAll(V... values) {
+	public void registerAll(@SuppressWarnings("unchecked") V... values) {
 		for (V value : values) {
 			register(value);
 		}
 	}
 
 	@Override
 	public boolean containsKey(Identifier key) {
-		return vanilla.containsId(key);
+		while (key != null) {
+			if (this.names.containsKey(key)) {
+				return true;
+			}
+
+			key = this.aliases.get(key);
+		}
+
+		return false;
 	}
 
 	@Override
 	public boolean containsValue(V value) {
-		return vanilla.getId(value) != null;
+		return this.names.containsValue(value);
 	}
 
 	@Override
 	public boolean isEmpty() {
-		return vanilla.getIds().isEmpty();
+		return this.names.isEmpty();
 	}
 
 	@Override
 	public V getValue(Identifier key) {
-		return vanilla.get(key);
+		V ret = this.names.get(key);
+		key = this.aliases.get(key);
+
+		while (ret == null && key != null) {
+			ret = this.names.get(key);
+			key = this.aliases.get(key);
+		}
+
+		return ret == null ? this.defaultValue : ret;
 	}
 
 	@Override
 	public Identifier getKey(V value) {
-		return vanilla.getId(value);
+		Identifier ret = this.names.inverse().get(value);
+		return ret == null ? this.defaultKey : ret;
+	}
+
+	@Override
+	public Set<Identifier> getKeys() {
+		return Collections.unmodifiableSet(this.names.keySet());
+	}
+
+	@Nonnull
+	@Override
+	public Collection<V> getValues() {
+		return Collections.unmodifiableSet(this.names.values());
+	}
+
+	@Override
+	public Set<Entry<Identifier, V>> getEntries() {
+		return Collections.unmodifiableSet(this.names.entrySet());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public <T> T getSlaveMap(Identifier name, Class<T> type) {
+		return (T) this.slaves.get(name);
 	}
 
+	@SuppressWarnings("unchecked")
 	@Override
+	public void setSlaveMap(Identifier name, Object obj) {
+		((Map<Identifier, Object>) this.slaves).put(name, obj);
+	}
+
+	public int getID(V value) {
+		Integer ret = this.ids.inverse().get(value);
+
+		if (ret == null && this.defaultValue != null) {
+			ret = this.ids.inverse().get(this.defaultValue);
+		}
+
+		return ret == null ? -1 : ret.intValue();
+	}
+
+	public int getID(Identifier name) {
+		return getID(this.names.get(name));
+	}
+
+	private int getIDRaw(V value) {
+		Integer ret = this.ids.inverse().get(value);
+		return ret == null ? -1 : ret.intValue();
+	}
+
+	private int getIDRaw(Identifier name) {
+		return getIDRaw(this.names.get(name));
+	}
+
+	public V getValue(int id) {
+		V ret = this.ids.get(id);
+		return ret == null ? this.defaultValue : ret;
+	}
+
+	void validateKey() {
```

Comment:
What's up with these package-private members? If they're unused they can safely be removed.
It would also reduce the noise of this PR a fair bit.